File: Medium/3016. Minimum Number of Pushes to Type Word II.java
import java.util.Arrays;

class Solution {
    public int minimumPushes(String word) {
        final var pushs = new int[] { 4, 4, 3, 3, 3, 3, 3, 3, 3, 3, 2, 2, 2, 2, 2, 2, 2, 2, 1, 1, 1, 1, 1, 1, 1, 1 };
        final var cnt = new int[26];
        for (var c : word.toCharArray()) {
            cnt[c - 'a']++;
        }
        Arrays.sort(cnt);
        int res = 0;
        for (int i = 25; i >= 0; i--) {
            res += pushs[i] * cnt[i];
        }
        return res;
    }
}

File: Medium/3016. Minimum Number of Pushes to Type Word II.py
class Solution:
    def minimumPushes(self, word: str) -> int:
        counter = Counter(word)
        sorted_freq = sorted([v for _, v in counter.items()], reverse = True)
        # print(sorted_freq)
        res = 0
        curr = 1
        left = 8
        for freq in sorted_freq:
            res += freq * curr
            left -= 1
            if not left:
                left = 8
                curr += 1
        return res
        

# wrong
class SolutionWrong:
    def minimumPushes(self, word: str) -> int:
        num_to_pushes = {}
        curr = 1
        left = 8
        min_pushes = 0
        for c in word:
            if c not in num_to_pushes:
                num_to_pushes[c] = curr
                left -= 1
                if left == 0:
                    left = 8
                    curr += 1
            min_pushes += num_to_pushes[c]
        return min_pushes
 